### DISCUSSION TITLE
[TCP] Connection fails to TCP server opened on Android device

### DISCUSSION THREAD

**Discussion by AndriiZozulia** [alec1o/Netly#33] [Q&A (Question and Answer)] [answered]
Hi!
I'm struggling with some weird behaviour of tcp server in my Unity project.
Basically I'm spawning server like that:
 ```
m_server = new TcpServer(false);
                m_server.OnOpen(() => LogUtils.Log("TCPServer", string.Format("TCP server instantiated on ip {0} and socket {1} on hostname {2}", m_serverHostAddress.ToString(), serverSetup.socket.ToString(), Dns.GetHostName())));
                m_server.OnEnter((client) =>
                {
                    LogUtils.Log("TCPServer", string.Format("TCP server received client"));
                });
                
                m_server.OnData((client, data) =>
                {
                    m_receivedData += Encoding.ASCII.GetString(data);
                });

                m_server.OnError((e) =>
                {
                    LogUtils.Halt("TCPServer", string.Format("TCP server error {0}", e.ToString()));
                });

                m_server.OnClose(() =>
                {
                    LogUtils.Log("TCPServer", string.Format("TCP server closed"));
                });
                
                m_server.Open(m_serverHost);
```

Currect implementation works very well when it's running in Unity Editor, but on Android device sokect is unavaliable though I'm getting OnOpen callback. Am I doing something wrong?
King regards.

**Comment by alec1o** [comment]
How can you know the socket is "unavaliable" on mobile device?

**Comment by alec1o** [comment]
Can you send me the log or more data about the error?

**Reply by AndriiZozulia** [reply]
I'm connecting to tcp sokect via Test-NetConnection

**Reply by alec1o** [reply]
What is your Host instance value? ``Netly.Core.Host``

**Reply by AndriiZozulia** [reply]
0,000000 [TCPServer]: TCP server instantiated on ip 169.254.2.44 and socket 8800 on hostname localhost

**Reply by alec1o** [reply]
OK! Lets find problem now.

* Create a ``TcpClient`` on your Unity project and try connect to the server.

> For we know if problem is socket on android.

* If the internal ``TcpClient`` connected successful on ``TcpServer`` it mean the problem isnot socket creation. (Maybe other things like IP)

**Reply by AndriiZozulia** [reply]
So I tested your proposal and socket is still unavaliable on Android and works well in Editor. Here is some listing and result that I get.
 ```
 void InstantiateServer()
    {
        ResolveServerAddress();
        if (m_serverAddressResolved)
        {
            try
            {
                m_server = new TcpServer(false);
                m_server.OnOpen(OnServerOpen);
                m_server.OnEnter((client) =>
                {
                    LogUtils.Log("TCPServer", string.Format("TCP server received client"));
                });

                m_server.OnData((client, data) =>
                {
                    m_receivedData += Encoding.ASCII.GetString(data);
                });

                m_server.OnError((e) =>
                {
                    LogUtils.Halt("TCPServer", string.Format("TCP server error {0}", e.ToString()));
                });

                m_server.OnClose(() =>
                {
                    LogUtils.Log("TCPServer", string.Format("TCP server closed"));
                });

                m_server.Open(m_serverHost);
            }
            catch (Exception exception)
            {
                LogUtils.Warning("TCPServer", "failed to handle tcp server instantiation. Check build and retry manualy. EXP: " + exception.ToString());
            }
        }
        else
        {
            LogUtils.Warning("TCPServer", "failed to resolve tcp server ip. Please, retry manualy.");
        }
    }

    protected void OnServerOpen()
    {
        LogUtils.Log("TCPServer", string.Format("TCP server instantiated on ip {0} and socket {1} on hostname {2}", m_serverHostAddress.ToString(), serverSetup.socket.ToString(), Dns.GetHostName()));
        
        var testInternalClient = new TcpClient(false);
        testInternalClient.OnOpen(() =>
        {
            MainThread.Add(() =>
            {
                LogUtils.Log("TCPServer", string.Format("TCP test internal client opened"));
                testInternalClient.ToData("test/test");
                testInternalClient.Close();
            });
        });

        testInternalClient.OnError((e) =>
        {
            LogUtils.Warning("TCPClient", "tcp test internal client instantiation failed: " + e.ToString());
            InstantiateClient();
        });


        testInternalClient.OnClose(() =>
        {
            MainThread.Add(() =>
            {
                LogUtils.Log("TCPServer", string.Format("TCP test internal client closed"));
            });
        });

        // open connection
        testInternalClient.Open(m_serverHost);
    }
```

![image](https://github.com/alec1o/Netly/assets/40180614/e98cbd0f-72f1-4113-8a51-144442e85f51)

And how it works in Editor
![image](https://github.com/alec1o/Netly/assets/40180614/3bf1bf4a-5162-4627-8f41-9fef1234dc92)

**Comment by alec1o** [comment]
Write this example on your Unity project.

> ```ConnectionExample.cs```

```cs
using System;
using Netly;
using Netly.Core;
using UnityEngine;

public class ConnectionExample : MonoBehaviour
{
    static void Log(string message)
    {
       	Debug.Log(message);
	return;
        // or
        // LogUtils.Log(message);
        // or
        // Whatever
        throw new NotImplementedException(nameof(Log));
    }

    internal TcpClient m_client = new TcpClient(false);
    internal TcpServer m_server = new TcpServer(false);
    internal Host m_host = new Host("0.0.0.0", PORT);
    internal const int PORT = 12000;
    internal int IDs = 0;

    private void Start()
    {
        Log("Starting Server...");

        m_server.OnClose(() => Log("[SERVER] Disconnected."));
        m_server.OnError((e) => Log($"[SERVER] Error on connect: {e}"));

        m_server.OnOpen(() =>
        {
            Log("[SERVER] Connected.");
            StartClient();
        });

        m_server.OnEnter(client =>
        {
            IDs++;
            int id = IDs;

            Log($"[SERVER] Client connected ({client.Host}) ID: {id}.");

            client.OnClose(() =>
            {
                Log($"[SERVER] Client disconnected ({client.Host} ID: {id})");
            });

            client.OnData((data) =>
            {
                Log($"[SERVER] Client received ({client.Host}) ID: {id}) DATA: {NE.GetString(data, NE.Mode.UTF8)}");
                
                // ECHO effect.
                client.ToData(data);
            });
        });

        m_server.Open(m_host);

        void StartClient()
        {
	    // host listen on "0.0.0.0", internal client connect on "127.0.0.1"
	    var host = new Host("127.0.0.1", m_host.Port);

            Log("Starting Client...");

            m_client.OnOpen(() =>
            {
                Log($"[CLIENT] Connected. From ({m_client.Host})");
                m_client.ToData(NE.GetBytes("Hello Server", NE.Mode.UTF8));
            });

            m_client.OnClose(() =>
            {
                Log($"[CLIENT] Disconnected. From ({m_client.Host})");
            });

            m_client.OnError((e) => Log($"[CLIENT] Error on connection: {e}"));

            m_client.OnData((data) => Log($"[CLIENT] Data: {NE.GetString(data, NE.Mode.UTF8)}"));

            m_client.Open(host);
        }
    }
}
```

> ```Output```
```log
Starting Server...
[SERVER] Connected.
Starting Client...
[CLIENT] Connected. From (127.0.0.1:12000)
[SERVER] Client connected (127.0.0.1:62548) ID: 1.
```

**Reply by AndriiZozulia** [reply]
Will try asap.

**Reply by AndriiZozulia** [reply]
![image](https://github.com/alec1o/Netly/assets/40180614/8f6c3b39-a873-4c40-9d3b-68bcc61a989f)

And we're still here.

**Reply by AndriiZozulia** [reply]
But it's nice working in Editor
![image](https://github.com/alec1o/Netly/assets/40180614/20f883d9-641f-4770-827a-a3682676ac6e)

**Reply by alec1o** [reply]
Can you give me complet log of ``[CLIENT] Error on connection: ...``

**Reply by AndriiZozulia** [reply]
Sorry about screenshots
![image](https://github.com/alec1o/Netly/assets/40180614/7a998ac1-42db-46fe-886a-96a0b7875765)

**Reply by alec1o** [reply]
Well. What is Device that are getting this error ``Android`` or ``Window Build``?

**Comment by alec1o** [comment]
Try put on ``m_host`` value as ``127.0.0.1`` instead of ``0.0.0.0`` (maybe ``0.0.0.0`` is unvalid server ip or whatever...) if not working it mean the problem is internal of ``OS Policy`` or ``System.Sockets.* Impl``  i wished get same environment that you to me test and try solve this problem. Server it's showing working but i don't know if the problem is ``OS`` blocking socket or the ``lib`` is broke.

* Can you give me details about build application target? ``(it will help me a lot)``

**Reply by AndriiZozulia** [reply]
Still actual?

**Reply by alec1o** [reply]
It's internal socket problem. maybe ``OS`` blocking socket or Socket failed on connect or listen on localhost

**Comment by alec1o** [comment]
After many test i find it.






> Runnig on Android 11 (same error) ``Real Device``
> Runnig on Android 9 (Android Emulator) ``MeMu Play``
> Runnig on Android 9 (Android Emulator) ``BlueStacks 5``

<img src="https://github.com/alec1o/Netly/assets/100610503/4abc6a99-1720-441c-8d72-d5f2becae8af" width="32%"/>
<img src="https://github.com/alec1o/Netly/assets/100610503/bea6e53f-6c26-4331-93fe-9a5431eb32c1" width="32%"/>
<img src="https://github.com/alec1o/Netly/assets/100610503/b63d8810-8af8-4417-9796-8e62b31a0475" width="32%"/>



<br/>

> ### The origin of error is internal and unknow

It's definitely a DOTNET issue on android. Netly uses ``System.Net.Sockets`` as the basis for socket creation. This library is having internal errors in the Android device.

My suspicions:
1. Android Manifest needs some kind of permissions.
2. The compilation of ``System.Net.Sockets`` internally is not working due to some internal type of conversion (C# to C++/Java or object).

Doubts:
1. The problem is with the TCP server or TcpClient. What about the server that is not working (listening directly?) or the TcpClient that cannot connect to the socket? (Needs to be tested)

2. In case the problem is the TcpServer that cannot listen to the localhost on Android or the TcpClient that cannot connect to the localhost. How should you solve it if the problem is internal? (Do nothing?)

3. How to move forward? Continue writing socket using ``System.Net.*``? or start the base in C++ and the interface in C#?

**Reply by AndriiZozulia** [reply]
So for now there is no solutuion that can fix current problem I guess?
Would you try to fix it in next releases? How long it will take?
Thank you for your fast replies and keeping in touch.

**Reply by alec1o** [reply]
1. I started try fix it now!
2. Anyway Netly 4.0 will released soon i wish fix it before Netly 4.0 be released.
3. I don't know real problem i need search and test it a lot to be find and solve the problem. because that i don't know how much time it need.

> If i get solve it you can use latest .dll and you can use fixed version before Netly 4.0 complete released.
I will tell you when i fix it. trust me i will do my best to fix it soon 🙏 .

**Reply by AndriiZozulia** [reply]
You awesome! Wish you good luck!
Hope to get back from you soon *handshake emoji*

**Reply by alec1o** [reply]
Fixed:  https://github.com/alec1o/Netly/discussions/33#discussioncomment-8407986

**Comment by alec1o** [comment, marked as answer]
``💯 Fixed``

Well, I started experimenting to find out if the problem was on the client or the server. By downloading the TCP debug application on GooglePlay, you can see that my client connected to the server created by the debuger app, but the debuger client did not connect to my server. 

Then I went and tried to create the server using TcpListener and it worked, as ``Netly`` and ``TcpListener`` use ``Sockets`` as a base, so I opened the source code of ``TcpListener`` and started to see the Possible modification and I entered the problem (``Backlog``).

(``Backlog``) Netly used the value (``0``) until it was corrected (db1eda4de5b891b83578a676f11a9e0b197756ff) and on Android (``backlog=zero``) is (``zero``) connections, and on Windows, value (``zero``) is the same as the default value. I had to leave this explicitly exposed in the code for this to be resolved.

### How to correct?
You can correct it if you need to use the development version, just use it.
> This step is necessary for versions lower than ``4.x.x`` and not necessary for higher versions.
```cs
// get default backlog value.
const int defaultBacklog = (int)System.Net.Sockets.SocketOptionName.MaxConnections;
// set backlog value on <Tcp.Server.Open> as param.
TcpServer.Open(<Netly.Core.Host> host, backlog: defaultBacklog );
```
![Screenshot (64)](https://github.com/alec1o/Netly/assets/100610503/8549c365-62d7-4128-aecc-32febec82141)

**Reply by AndriiZozulia** [reply]
Works well for me. Thank you so much for fast iterations!

**Reply by alec1o** [reply]
🔥 Thanks for improve Netly to be better by solving bugs!

**Comment by pequi-roido** [comment]
Hi, glad it will get fixed in the next version! (and thanks for this incredible library!)

Just adding a bit of information on this issue, in my app using NET8 on Android, the issue only appeared in devices running Android 13 and up (gladly, with the backlog fix everything is fine now!). 

On Android 12 and lower, weirdly, the code can run unaltered.

**Reply by alec1o** [reply]
Thanks 💕 for the compliments.
And I hope this library continues to help. 🔥 


_The reason for running on old Android devices is that the previous api accepted 0 backlog as unlimited backlogs and in the modern api 0 backlog is connection number would be "accepted"_
